vm/src/stdlib/grp.rs

```diff
@@ -0,0 +1,113 @@
+pub(crate) use grp::make_module;
+
+#[pymodule]
+mod grp {
+    use std::ptr::NonNull;
+
+    use crate::{
+        builtins::{PyIntRef, PyListRef, PyStrRef},
+        convert::{IntoPyException, ToPyObject},
+        exceptions,
+        types::PyStructSequence,
+        AsObject, PyObjectRef, PyResult, VirtualMachine,
+    };
+    use nix::unistd;
+
+    #[pyattr]
+    #[pyclass(module = "grp", name = "struct_group")]
+    #[derive(PyStructSequence)]
+    struct Group {
+        gr_name: String,
+        gr_passwd: String,
+        gr_gid: u32,
+        gr_mem: PyListRef,
+    }
+    #[pyimpl(with(PyStructSequence))]
+    impl Group {
+        fn from_unistd_group(group: unistd::Group, vm: &VirtualMachine) -> Self {
+            let cstr_lossy = |s: std::ffi::CString| {
+                s.into_string()
+                    .unwrap_or_else(|e| e.into_cstring().to_string_lossy().into_owned())
+            };
+            Group {
+                gr_name: group.name,
+                gr_passwd: cstr_lossy(group.passwd),
+                gr_gid: group.gid.as_raw(),
+                gr_mem: vm
+                    .ctx
+                    .new_list(group.mem.iter().map(|s| s.to_pyobject(vm)).collect()),
+            }
+        }
+    }
+
+    // impl From<unistd::Group> for Group {
+    //     fn from(group: unistd::Group) -> Self {
+    //         // this is just a pain...
+    //         let cstr_lossy = |s: std::ffi::CString| {
+    //             s.into_string()
+    //                 .unwrap_or_else(|e| e.into_cstring().to_string_lossy().into_owned())
+    //         };
+    //         Group {
+    //             gr_name: group.name,
+    //             gr_passwd: cstr_lossy(group.passwd),
+    //             gr_gid: group.gid.as_raw(),
+    //             gr_mem: group.mem,
+    //         }
+    //     }
+    // }
+
+    #[pyfunction]
+    fn getgrgid(gid: PyIntRef, vm: &VirtualMachine) -> PyResult<Group> {
+        let gid_t = libc::gid_t::try_from(gid.as_bigint()).map(unistd::Gid::from_raw);
+        let group = match gid_t {
+            Ok(gid) => unistd::Group::from_gid(gid).map_err(|err| err.into_pyexception(vm))?,
+            Err(_) => None,
+        };
```

Comment:
```suggestion
        let gid = libc::gid_t::try_from(gid.as_bigint()).map(unistd::Gid::from_raw);
        let group = gid.ok().map(|gid| unistd::Group::from_gid(gid)).transpose()?;
```

Reply:
If the following is performed, the type of the result is Option<Option<Group> instead of Option<Group>.

Reply:
Thank you, I have read the material you sent.
and_then() seems to receive a function as an argument, but shouldn't it be implemented so that option<option<string>> can be made into option<string> in that function?

Reply:
I think @MrGreenTea meant `flatten`.

https://doc.rust-lang.org/std/option/enum.Option.html#method.flatten

Reply:
I'm not sure how, but why not just do the unwrap() twice?
`let group = gid.ok().map(|gid| unistd::Group::from_gid(gid)).transpose().unwrap().unwrap();`

Reply:
Oh, I didn't catch git_t was used later. `gid_t` is a type name. Could you give it another proper name?

Always avoid unwrap unless it never happens. We are making interpreter. Our users want to see Python exceptions instead of Rust error when they failed but `unwrap()` panics in Rust stack.

Reply:
Since gid should be used later and gid_t is created only for .ok, I think it would be better to use it without making it a variable.

So, how to safely unwrap an option<option<string>> object into an option<string>?
`let group = gid_t.ok().map(|gid| unistd::Group::from_gid(gid)).transpose()?.flatten();`
If you use something like this, the problem with ?

Reply:
I believe it could be

```rs
gid_t.ok().and_then(|gid| unistd::Group::from_gid(gid)).transpose()?
```

Reply:
Well, I didn't look in the code carefully when I left last comment. `gid_t` can be safely replaced by `gid` because it is never used after its last usage to create `git`. You can reuse same variable names again and again if you don't need the previous names later.

Line 50 expect `group` to be `Option`. `transpose()?.flatten()` will work.

Reply:
@MrGreenTea That doesn't looking work because one is Option and the other one is Result.

Reply:
If in doubt, follow the types.

```rust
let group = gid_t                                         // Result<T, E>
                 .ok()                                    // Option<T>
                 .map(|gid| unistd::Group::from_gid(gid)) // Option<Result<Option<Group>>>
                 .transpose()                             // Result<Option<Option<Group>>, nix::errno::Errno>
                 ?                                        // Option<Option<Group>>
                 .flatten();                              // Option<Group>
```

Does RustPython have a way to convert `nix::errno::Errno` into some sort of Exception?

Reply:
I understand what you said.
However, it is not possible to use gid as a variable name because it has a gid in the argument in the same form as getgrgid(gid, vm) and needs to be used later.

And since gid_t will not be used later, I think it will be okay to use it without declaring it as a variable.

Reply:
@fanninpm In the code you mentioned, the following error occurs in `?`
`grp.rs(44, 56): expected `core::PyRef<exceptions::types::PyBaseException>` because of this`

Reply:
You would need a call to `map_err()` between `transpose()` and `?` to convert `nix::errno::Errno` into something compatible with `PyBaseException`.

Reply:
@fanninpm I was wondering if the code related to the exception disappeared as the code was changed, but this is the problem! Thanks it has been fixed thanks!